8Knot/pages/index/search_utils.py

```diff
@@ -0,0 +1,248 @@
+    """
+    Calculate a similarity score based on Levenshtein distance.
+    
+    Args:
+        s1: First string (typically the search query)
+        s2: Second string (typically the option label)
+        
+    Returns:
+        float: Score between 0 and 1, where 1 is a perfect match
+    """
+    # Calculate the Levenshtein distance
+    distance = levenshtein_distance(s1, s2)
+    
+    # Get the maximum possible distance (which is the length of the longer string)
+    max_distance = max(len(s1), len(s2))
+    
+    # If both strings are empty, they are identical
+    if max_distance == 0:
+        return 1.0
+    
+    # Calculate the similarity score (1 - normalized distance)
+    # The closer to 0 the distance is, the closer to 1 the score will be
+    return 1.0 - (distance / max_distance)
+
+def simple_fuzzy_score(s1: str, s2: str) -> float:
+    # Convert both strings to lowercase for case-insensitive matching
+    s1, s2 = s1.lower(), s2.lower()
+    
+    # Check for exact substring match first (give it a high score)
+    if s1 in s2:
+        # If it's at the beginning of the string, give it an even higher score
+        if s2.startswith(s1):
+            return 1.0
+        return 0.9
+    
+    # Use sequence matcher for fuzzy matching
+    return SequenceMatcher(None, s1, s2).ratio()
+
+def tokenize_search(query: str) -> List[str]:
+    """
+    Split search query into tokens for more flexible matching.
+    
+    Args:
+        query: Search query string
+        
+    Returns:
+        List of tokens from the query
+    """
+    return [token.strip().lower() for token in re.split(r'[\s/_.,-]+', query) if token.strip()]
+
+def token_match_score(tokens: List[str], label: str) -> float:
+    """
+    Calculate match score based on tokens using Levenshtein distance.
+    
+    Args:
+        tokens: List of search tokens
+        label: String to match against
+        
+    Returns:
+        float: Score between 0 and 1
+    """
+    if not tokens:
+        return 0.0
+        
+    label_lower = label.lower()
+    label_tokens = tokenize_search(label_lower)
+    
+    # Check for token matches with Levenshtein distance
+    match_scores = []
+    for token in tokens:
+        # Try exact contains first
+        if token in label_lower:
+            match_scores.append(0.95)
+            continue
+            
+        # If no exact match, calculate best score against label tokens
+        best_score = 0.0
+        for label_token in label_tokens:
+            # Use Levenshtein for better matching with typos
+            score = levenshtein_score(token, label_token)
+            best_score = max(best_score, score)
+        
+        match_scores.append(best_score)
+    
+    # Average the scores
+    if match_scores:
+        return sum(match_scores) / len(match_scores)
+    return 0.0
+
+def fuzzy_search(query: str, options: List[Dict[str, Any]], threshold: float = 0.2) -> List[Dict[str, Any]]:
```

Comment:
consider replacing all of this and using FuzzyWuzzy:
"FuzzyWuzzy uses the Levenshtein distance implementation of fuzzy matching to give you a ratio score between the provided sentences." we use fuzzywuzzy in some of out visualization preprocessing